### PR TITLE
Moneo Shamir

### PR DESCRIPTION
**Pull request by Sheepe**
A datapack expansion to metallurgy adding a new shamir, Moneo.

This shamir idea was created by InternetAlien and developed by myself (SirSheepe).

This shamir can be applied to all `wood`, `stone`, `iron`, `diamond` and `gold` tools, as well as `shears`. It will give the player mining fatigue when the durability of the tool they are using is low, which is different for each material. There are two of these "warning" levels, each one comes with a sound and a durability range.

Wood:
    Level 1: 20-11 durability
    Level 2: Less than or equal to 10 durability

Stone:
    Level 1: 30-21 durability
    Level 2: Less than or equal to 20 durability

Iron:
    Level 1: 40-21 durability
    Level 2: Less than or equal to 20 durability

Diamond:
    Level 1: 50-21 durability
    Level 2: Less than or equal to 20 durability

Gold:
    Level 1: 10-6 durability
    Level 2: Less than or equal to 5 durability

Shears:
    Level 1: 35-21 durability
    Level 2: Less than or equal to 20 durability

The user can still break the tool if they continue mining past this point.

A high pitched sound will play for level 1, and a lower pitched sound will play for level 2. The sound will play whilst the user is holding their tool.

The shamir is made with ~~Thorium Brass~~ Aluminium.

----

**This has been resolved**

> To preserve compatibility to add new shamirs in the future, I found that it will likely be best that all metallurgy expansions have a tag which runs their individual "main" functions, hence this shamir will only work if `#metallurgy_expansion:main` is being called in the `metallurgy:main` function in the base datapack. If you have a better solution to this problem, I'd like to hear it.

### CODE REVIEW

**Comment by Bluefire-610** [comment]
I'm all in on this one! Nice idea and possibly really useful. Though I'm not sure about adding a function tag for the shamir's main.
The only reason why shamirs have to interface with metallurgy base at all is for the crafting mechanic (both smelting and putting the shamir on a tool). Think of it like heart canisters: They need custom crafters to be created, but not to function.

Therefore a main function tag does not exist. The shamir's functionality should be handled by it's own main, called by gm4's clocking system.

File: gm4_metallurgy_moneo/pack.mcmeta [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+  "pack":{
+    "pack_format":0,
+    "description":"A Gamemode 4 Module"
+  },
+  "module_name":"Metallurgy Moneo",
+  "module_id":"metallurgy_moneo",
```
Comment (misode) [inline gm4_metallurgy_moneo/pack.mcmeta]: This should be `Moneo Shamir` and `moneo_shamir` imo (also applies in `init.mcfunction`), but feel free to discuss this as there haven't been any shamir expansions yet.

The translation key should then also be `module.gm4.moneo_shamir`

Comment (Sheepe) [reply gm4_metallurgy_moneo/pack.mcmeta]: Alright that seems reasonable, I'll make the requested changes.

Comment (Sheepe) [reply gm4_metallurgy_moneo/pack.mcmeta]: What about the installation text? It currently says `Metallurgy Expansion (Moneo)`. I'm assuming that should just be `Moneo Shamir`?

Comment (Bluefire-610) [reply gm4_metallurgy_moneo/pack.mcmeta]: Yup! It's basically a standalone module feature wise, just using metallurgy for crafting.

**Comment by sparks0** [comment]
Looks really cool!  This should go on Aluminium bands though as I would count this as "Supression"

**Comment by Kroppeb** [comment]
What metal has what ability again? I can’t seem to find it on the wiki.

**Comment by sparks0** [comment]
> What metal has what ability again? I can’t seem to find it on the wiki.

I did say Aluminium :) I'm updating the wiki page right now to include a metal property table since that would be useufl on there.

**Comment by Kroppeb** [comment]
Barium:
> increases player awareness

Aluminium:
> limits the player's awareness e.g. mining fatique or slowness.

**Comment by Bluefire-610** [comment]
image
@Kroppeb

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#run from main
```
Comment (Kroppeb) [inline gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: These are 14 nbt serialisations in one file

**Comment by Sheepe** [comment]
Hey, I hear ya. I can change it to the Aluminium shamir, however the reason it doesn't support Gold tools is their ridiculously low durability. Though, since you insist, I'll add them. If a better sound is able to be given, then I'll gladly implement.

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction (hunk 2 above)
Comment (Sheepe) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: If you know a better way to check for the tool material and execute the function, I'd like to know lol

Comment (misode) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: @Sheepe Simple. You store the material max when the shamir is applied once. Then you don't need to branch for each item id, just do the operation.

Comment (Sheepe) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: Like, store the material maximum inside of the NBT?

Comment (Sheepe) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: Or do you mean, give the player a score of the maximum, meaning it's only recalculated when they start holding the tool.

Comment (Sheepe) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: Either way, I think I did what you suggested and the amount of checks should be greatly reduced.

Comment (misode) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: I actually meant to store it on the tool, but this looks good as well

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction [hunk 3]
```diff
@@ -1,27 +1,19 @@
...
-execute if entity @s[nbt={SelectedItem:{id:"minecraft:stone_shovel"}}] run function moneo_shamir:tools/materials/stone
-execute if entity @s[nbt={SelectedItem:{id:"minecraft:stone_pickaxe"}}] run function moneo_shamir:tools/materials/stone
-execute if entity @s[nbt={SelectedItem:{id:"minecraft:stone_hoe"}}] run function moneo_shamir:tools/materials/stone
-execute if entity @s[nbt={SelectedItem:{id:"minecraft:stone_axe"}}] run function moneo_shamir:tools/materials/stone
-
-execute if entity @s[nbt={SelectedItem:{id:"minecraft:iron_shovel"}}] run function moneo_shamir:tools/materials/iron
-execute if entity @s[nbt={SelectedItem:{id:"minecraft:iron_pickaxe"}}] run function moneo_shamir:tools/materials/iron
-execute if entity @s[nbt={SelectedItem:{id:"minecraft:iron_hoe"}}] run function moneo_shamir:tools/materials/iron
-execute if entity @s[nbt={SelectedItem:{id:"minecraft:iron_axe"}}] run function moneo_shamir:tools/materials/iron
+execute unless score @s gm4_moneo_max matches 0.. run function moneo_shamir:store_maximum
```
Comment (Kroppeb) [inline gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: Wouldn’t this have the wrong max if you switch moneo tools?

Comment (misode) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: Oh right, good point

Comment (Sheepe) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: I guess that's what I get for working on it at 2-3am

Comment (Sheepe) [reply gm4_metallurgy_moneo/data/moneo_shamir/functions/active.mcfunction]: This should now be resolved.

**Comment by Bluefire-610** [comment]
Moneo seems to never be marked as installed.
Anyone else get the install message on every reload?

**Comment by Sheepe** [comment]
That was fixed in https://github.com/Gamemode4Dev/GM4_Datapacks/pull/259/commits/9118fc0d31c437e1975cc9b053ed6f02731aa56a

**Comment by Bluefire-610** [comment]
Waiting on a change with the playsound, we think a metal-like sound would be better.

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/diamond.mcfunction [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# run from active
+# @s = player holding the moneo tool
+
+execute if score tool_max_damage gm4_ml_data matches 21..50 run effect give @s minecraft:mining_fatigue 2 0
+execute if score tool_max_damage gm4_ml_data matches ..50 at @s run playsound minecraft:block.lantern.hit master @s ~ ~ ~ 5 2
```
Comment (Bluefire-610) [inline gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/diamond.mcfunction]:
```suggestion
execute if score tool_max_damage gm4_ml_data matches ..50 at @s run playsound minecraft:block.lantern.hit player @s ~ ~ ~ 5 2
```

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/gold.mcfunction [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# run from active
+# @s = player holding the moneo tool
+
+execute if score tool_max_damage gm4_ml_data matches 6..10 run effect give @s minecraft:mining_fatigue 2 0
+execute if score tool_max_damage gm4_ml_data matches ..10 at @s run playsound minecraft:block.lantern.hit master @s ~ ~ ~ 5 1.4
```
Comment (Bluefire-610) [inline gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/gold.mcfunction]:
```suggestion
execute if score tool_max_damage gm4_ml_data matches ..10 at @s run playsound minecraft:block.lantern.hit player @s ~ ~ ~ 5 1.4
```

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/iron.mcfunction [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# run from active
+# @s = player holding the moneo tool
+
+execute if score tool_max_damage gm4_ml_data matches 21..40 run effect give @s minecraft:mining_fatigue 2 0
+execute if score tool_max_damage gm4_ml_data matches ..40 at @s run playsound minecraft:block.lantern.hit master @s ~ ~ ~ 5 1
```
Comment (Bluefire-610) [inline gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/iron.mcfunction]:
```suggestion
execute if score tool_max_damage gm4_ml_data matches ..40 at @s run playsound minecraft:block.lantern.hit player @s ~ ~ ~ 5 1
```

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/stone.mcfunction [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# run from active
+# @s = player holding the moneo tool
+
+execute if score tool_max_damage gm4_ml_data matches 21..30 run effect give @s minecraft:mining_fatigue 2 0
+execute if score tool_max_damage gm4_ml_data matches ..30 at @s run playsound minecraft:block.lantern.hit master @s ~ ~ ~ 5 0.7
```
Comment (Bluefire-610) [inline gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/stone.mcfunction]:
```suggestion
execute if score tool_max_damage gm4_ml_data matches ..30 at @s run playsound minecraft:block.lantern.hit player @s ~ ~ ~ 5 0.7
```

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/wood.mcfunction [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# run from active
+# @s = player holding the moneo tool
+
+execute if score tool_max_damage gm4_ml_data matches 11..20 run effect give @s minecraft:mining_fatigue 2 0
+execute if score tool_max_damage gm4_ml_data matches ..20 at @s run playsound minecraft:block.lantern.hit master @s ~ ~ ~ 5 0
```
Comment (Bluefire-610) [inline gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/materials/wood.mcfunction]:
```suggestion
execute if score tool_max_damage gm4_ml_data matches ..20 at @s run playsound minecraft:block.lantern.hit player @s ~ ~ ~ 5 0
```

File: gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/shears.mcfunction [hunk 9]
```diff
@@ -0,0 +1,6 @@
+# run from active
+# @s = player holding the moneo tool
+
+execute if score tool_max_damage gm4_ml_data matches 21..35 run effect give @s minecraft:mining_fatigue 2 0
+execute if score tool_max_damage gm4_ml_data matches ..35 at @s run playsound minecraft:block.lantern.hit master @s ~ ~ ~ 5 1
```
Comment (Bluefire-610) [inline gm4_metallurgy_moneo/data/moneo_shamir/functions/tools/shears.mcfunction]:
```suggestion
execute if score tool_max_damage gm4_ml_data matches ..35 at @s run playsound minecraft:block.lantern.hit player @s ~ ~ ~ 5 1
```

**Comment by misode** [comment]
>  I think the sound should play on the player slider.

I think master is fine actually. It would be really annoying to miss the warning because you disabled player sounds. I personally have player sounds turned off.

**Comment by Sheepe** [comment]
Can we please resolve whether we want it on the player slider or not?

**Comment by sparks0** [comment]
I'm in favour of keeping it on Master based on Misode's testimony. I think players will want to hear the warning no matter what or what's the point of a warning?

**Comment by Bluefire-610** [comment]
Alright, I'm fine with that then. I guess the only downside is that players won't be able to individually adjust the volume of the sound - which could be an issue in noisy areas.

**Comment by Bluefire-610** [comment]
Just tested this in game again and it works as intended. Only thing I noticed is the name of the module folder. Currently it is `gm4_metallurgy_moneo`, should this be changed to `gm4_moneo` @sparks0 @SpecialBuilder32 ?

**Comment by Sheepe** [comment]
I think it should be `gm4_moneo_shamir` to fit with the installation name and such.

**Comment by Bluefire-610** [comment]
> I think it should be `gm4_moneo_shamir` to fit with the installation name and such.

Sounds like a good choice to me!

**Comment by Sheepe** [comment]
I have no idea how to change it and I'm on mobile until Saturday.

**Comment by misode** [comment]
@Sheepe I made a PR on your fork renaming the folder https://github.com/Sheepe/GM4_Datapacks/pull/1

**Pull request merged**